File: solution2.js
/*
 * Given an array of integers A with N items,
 * get the longest switching segment.
 * A switching segment is when consecutive numbers alternate between two values.
 * e.g. [1, 2, 1, 2, 1]
 * If an array of length 1 is given (e.g. [2]), the function should return 1.
 * */

function solution(A) {
    function solution_recurs(arr, longest, segLen) {
        if (arr.length < 3) return longest;
        
        if (arr[0] === arr[2]) {
            if (segLen < 3) segLen = 3;
            if (arr.length > 3 && arr[1] === arr[3]) {
                segLen += 1;
            }

            longest = Math.max(segLen, longest);
        } else {
            segLen = 1;
        }

        return solution_recurs(arr.slice(1), longest, segLen);
    }



    return solution_recurs(A, 1, 1);
}

File: solution2.py
def solution(A):
  def solution_recurs(arr, longest, seg_len):
    if len(arr) < 3:
      return longest

    if arr[0] == arr[2]:
      if seg_len < 3:
        seg_len = 3

      if (len(arr) > 3) and (arr[1] == arr[3]):
        seg_len = seg_len + 1

      longest = max(seg_len, longest)
    else:
      seg_len = 1

    return solution_recurs(arr[1:], longest, seg_len)

  return solution_recurs(A, 1, 1)
